File: proofload-java/src/main/java/io/github/matthewjones372/proofload/java/Searches.java
package io.github.matthewjones372.proofload.java;

import io.github.matthewjones372.proofload.Capacity;
import io.github.matthewjones372.proofload.Goal;
import io.github.matthewjones372.proofload.Rate;
import io.github.matthewjones372.proofload.Rung;
import io.github.matthewjones372.proofload.Scenario;
import io.github.matthewjones372.proofload.Search;
import java.time.Duration;
import java.util.List;

/**
 * The hunt for the highest rate a scenario sustains, and the rates its answer
 * is read off.
 *
 * Kotlin writes {@code checkout.sustainable(500.perSecond, 2.minutes, goals)}.
 * Both of those arguments are value classes and so is every rate the curve
 * carries, which is why these are Java sources over {@link Hunts}.
 *
 * {@code Proofload.create().run(search)} sends it; the search itself is a value
 * and runs nothing.
 */
public final class Searches {

    private Searches() {
    }

    public static Search sustainable(Scenario scenario, Rate upTo, Duration holding, List<Goal> expecting) {
        return Hunts.sustainable(scenario, upTo.getPerSecond(), holding, expecting);
    }

    /** The same search, with every rung warmed for {@code over} at that rung's own rate before it is measured. */
    public static Search warmingUp(Search search, Duration over) {
        return Hunts.warmingUp(search, over);
    }

    /** The highest rate every goal held at, or null where even the lowest rung missed one. */
    public static Rate rate(Capacity capacity) {
        return (Rate) Hunts.capacityRate(capacity);
    }

    public static Rate rate(Rung rung) {
        return (Rate) Hunts.rungRate(rung);
    }

    /** The rate the load actually left at, which is under {@link #rate(Rung)} where the injector fell behind. */
    public static Rate offered(Rung rung) {
        return (Rate) Hunts.offered(rung);
    }
}

File: proofload-scala/src/main/scala/io/github/matthewjones372/proofload/scala/Searches.scala
package io.github.matthewjones372.proofload.scala

import io.github.matthewjones372.proofload.Capacity
import io.github.matthewjones372.proofload.Goal
import io.github.matthewjones372.proofload.Rate
import io.github.matthewjones372.proofload.RunResult
import io.github.matthewjones372.proofload.Rung
import io.github.matthewjones372.proofload.Scenario
import io.github.matthewjones372.proofload.Search
import io.github.matthewjones372.proofload.Verdict
import io.github.matthewjones372.proofload.java.Searches
import _root_.scala.concurrent.duration.FiniteDuration
import _root_.scala.jdk.CollectionConverters.ListHasAsScala
import _root_.scala.jdk.CollectionConverters.SeqHasAsJava

extension (scenario: Scenario)

  /**
   * The highest rate this scenario sustains while every goal holds, hunted
   * below `upTo`. A vararg rather than a list, matching `scenario(...)`.
   */
  def sustainable(upTo: Rate, holding: FiniteDuration, expecting: Goal*): Search =
    Searches.sustainable(scenario, upTo, asJava(holding), expecting.toList.asJava)

extension (search: Search)

  /** The same search, with every rung warmed for `over` at that rung's own rate before it is measured. */
  def warmingUp(over: FiniteDuration): Search = Searches.warmingUp(search, asJava(over))

extension (capacity: Capacity)

  /** The highest rate every goal held at, empty where even the lowest rung missed one. */
  def rate: Option[Rate] = Option(Searches.rate(capacity))

  /** Every rung that ran, lowest first. */
  def curve: Seq[Rung] = capacity.getCurve.asScala.toSeq

  /** The goal that stopped the climb. */
  def limitedBy: Option[Goal] = Option(capacity.getLimitedBy)

  /** True where a rung ended void, which makes [[rate]] a floor the generator reached rather than the target's ceiling. */
  def voided: Boolean = capacity.getVoided

extension (rung: Rung)

  def rate: Rate = Searches.rate(rung)

  def result: RunResult = rung.getResult

  def verdicts: Seq[Verdict] = rung.getVerdicts.asScala.toSeq

  def outcome: Rung.Outcome = rung.getOutcome
